### ISSUE TITLE
Prisma doesn’t handle multiple requests well

### ISSUE THREAD

**Issue by elie222** [prisma/prisma#9569]
I've created a repo with a reproduction of our issue. See the readme for notes on how to run it:

https://github.com/elie222/prisma-connection-bug

We ran into an issue where we were making too many concurrent requests to the database so the inserts would fail. 

Ideally Prisma would handle this case for us so that when we make too many requests it waits to make more. There's an option to set connection limit but it doesn't seem to respect it as it will make more connections than the limit.

**Comment by elie222** [comment]
What's the best way to handle this from an application stand point? Should I be catching the Prisma errors and then backing off if we're being too aggressive? Is there an easy global solution to apply to all prisma actions? Using middleware perhaps?

**Comment by janpio** [comment]
> I had it set to 60s. I'm setting it to unlimited now. Will see how it goes.

Any update on this @elie222?

**Comment by elie222** [comment]
I actually don't remember now but I'm avoiding that situation for the most part. We basically await every db interaction now.
Expecting to wait more than 60s not a great idea ever really. We'd start waiting indefinitely at some point. We process a lot of data and it doesn't slow down ever.

**Comment by janpio** [comment]
Can we close this issue @elie222 or is there anything else we should still look into?

**Comment by elie222** [comment]
> Can we close this issue @elie222 or is there anything else we should still look into?

We still run into these issues when running many requests in parallel. I'm surprised more issues don't face these same issues. Some sort of queueing system would be ideal where Prisma doesn't try and make 100 updates in parallel if it doesn't have access to 100 connections because connection limit is 20.

**Comment by janpio** [comment]
> Some sort of queueing system would be ideal

That exists and sits in front of the connection pool. It has a timeout, and after that timeout is over it triggers the `Timed out fetching a new connection from the connection pool.` error message. So if you increase or remove the timeout, the application will queue requests longer or forever, and will just wait until a connection opens up to handle it.

Or is that different from what you want?

**Comment by elie222** [comment]
Got it. Getting some sort of idea how big that queue is might be helpful. Atm seems like we have to guess a bit

**Comment by janpio** [comment]
That is true, Prisma does not help much here. (We are working on parts of that under the "Tracing" project)
But generally, when you get the timeout error message, it is definitely to full.

**Comment by nilkothari** [comment]
@janpio can you please check #13250, I am kind of facing similar issue. It would be great if you can provide more information on that.
